Cassandra 16956 - remove Windows support

smiklosovic opened the pull request.

bowensong commented on bin/cqlsh.py:
```diff
@@ -93,11 +93,6 @@
 # is a ../lib dir, use bundled libs there preferentially.
 ZIPLIB_DIRS = [os.path.join(CASSANDRA_PATH, 'lib')]
 myplatform = platform.system()
```
After removing the `is_win` part, the `myplatform` variable is only being used at `if myplatform == 'Linux':` below. It makes more sense to remove the global variable and change the line to `if platform.system() == 'Linux':`.

bowensong commented on bin/cqlsh.py:
```diff
@@ -1894,7 +1878,7 @@ def do_clear(self, parsed):
         """
         import subprocess
-        subprocess.call(['clear', 'cls'][is_win], shell=True)
+        subprocess.call(['clear', 'cls'][False], shell=True)
```
That new code is equivalent to `subprocess.call('clear', shell=True)`, but less readable.

bowensong commented on pylib/cqlshlib/test/run_cqlsh.py:
```diff
@@ -126,21 +118,12 @@ def timing_out_alarm(seconds):
-        return eventlet.Timeout(seconds, TimeoutError)
+# setitimer is new in 2.6, but it's still worth supporting, for potentially
+# faster tests because of sub-second resolution on timeouts.
+if hasattr(signal, 'setitimer'):
```
`signal.setitimer()` is available on all Python versions (2.7 and 3) currently supported by cqlsh. The `if hasattr(...)` is redundant.

bowensong commented on pylib/cqlshlib/test/run_cqlsh.py:
```diff
@@ -297,11 +277,6 @@ def __init__(self, path=None, host=None, port=None, keyspace=None, cqlver=None,
-        if tty and is_win():
-            args += ('--tty',)
-            args += ('--encoding', 'utf-8')
-            if win_force_colors:
```
After removing this line, the `win_force_colors` parameter in the `__init__()` method is unused and should be removed too. The `test_no_color_output` test in the test_cqlsh_output.py file needs to be updated or removed.

bowensong commented on pylib/cqlshlib/test/run_cqlsh.py:
```diff
@@ -30,15 +30,8 @@
-    return sys.platform in ("cygwin", "win32")
-
-if is_win():
-    from .winpty import WinPty
```
`WinPty` is still being referenced in the `ProcRunner.start_proc()` method. Please double check and make sure all references to the `WinPty` module has been taken care of.

bowensong reviewed (commented):
bin/cqlsh.py

After removing the `check_windows_encoding()` method, the `is_using_utf8` property is no longer being referenced and should be removed too.

The `CP65001` variable is Windows specific, and after removing the `check_windows_encoding()` method and the `is_using_utf8` property, it is unreferenced and should be removed.

The module level `insert_driver_hooks()` call should be moved into the `main()` method to stop monkey patching the cassandra Python driver code at module import time. On Linux/Unix the multiprocessing (used by `COPY TO`/`COPY FROM`) module uses `os.fork()` (instead of a subprocess on Windows), so the patch made in the `main()` method will presisit.

---

bin/debug-cql and bin/cassandra

CYGWIN and MINGW are Windows specific and should be removed.

---

pylib/cqlshlib/test/run_cqlsh.py

The `ProcRunner.read_winpty()` method is Windows specific and should be removed.

---

This is all I can find for now, and I'm pretty sure I have missed some.

jmckenzie-dev commented on src/java/org/apache/cassandra/db/lifecycle/LogTransaction.java:
```diff
@@ -113,7 +112,6 @@ public CorruptTransactionLogException(String message, LogFile txnFile)
     private final Object lock;
     private final Ref<LogTransaction> selfRef;
     // Deleting sstables is tricky because the mmapping might not have been finalized yet,
```
nit: terminate with period.

jmckenzie-dev commented on src/java/org/apache/cassandra/utils/NativeLibrary.java:
```diff
@@ -143,10 +140,8 @@ private static OSType getOsType()
-            return WINDOWS;
 
-        logger.warn("the current operating system, {}, is unsupported by cassandra", osName);
+        logger.warn("the current operating system, {}, is unsupported by Cassandra", osName);
```
Have we checked to confirm that, on windows under WSL, osType comes up as LINUX?

smiklosovic replied:
I am not sure, I guess I have to check.

bowensong replied:
based on the information from https://stackoverflow.com/questions/66736736/what-is-the-os-name-system-property-value-in-java-on-windows-subsystem-for-linux I'm pretty sure the `osName` is going to be `"linux"`. Some other StackOverflow posts also supports that. For example: https://stackoverflow.com/questions/68228844/maven-build-failed-due-to-permission-denied-exception-during-test-phase

jmckenzie-dev reviewed (commented):
You can also:

- Remove the comment in CompactionTask.java:173 about SSTableScanners on Windows
- Remove the comment at CassandraOutgoingFile.java:158 about Windows lack of atomic replace in index summary redistribution
- Remove the comments in PathUtils about windows being picky
- Modify the comment in SchemaConstants.java to indicate that we *used* to care about Windows path length restrictions before we removed support
- Remove the entire comment with benchmark reference results in DirectorySizerBench.java
- Possibly remove the logic in LogbackStatusListener.java:124 where we try and filter out the windows newline? (can't imagine how we'd get one on linux or osx but maybe we keep this just in case...)
- Remove the comment constraining impl behavior in ServerTestUtils:139 where we discuss needing to stop and unmap prior to cleanup() due to windows reqs
- Remove the comment in SystemKeyspaceTest.java about windows being unable to delete hard-links
- Remove the comment in LogTransactionTest.java about needing to use File.exists on windows
- Consider removing the System.gc() call and comment in SSTableLoaderTest.java which was added during CASSANDRA-16235 to deal with us missing buffer cleanup due to windows env. Or just remove the comment and keep the GC; that'd probably be fine too.

In general, all of these actions ☝️ will help unconstrained future maintainers from implementation details, refactoring, or cleanups they can engage in or potential future confusion about whether or not we need to keep things a certain way for certain environments.

jmckenzie-dev commented on doc/modules/cassandra/pages/troubleshooting/index.adoc:
```diff
@@ -8,10 +8,7 @@ isolate the problem using logs and tools. Luckily Cassandra had a great
 
 These pages include a number of command examples demonstrating various
-debugging and analysis techniques, mostly for Linux/Unix systems. If you
-don't have access to the machines running Cassandra, or are running on
```
Would keep the bit about "If you don't have access to the machines running C*" etc etc

jmckenzie-dev commented on pylib/cqlshlib/test/run_cqlsh.py:
```diff
@@ -111,36 +103,6 @@ def timing_out_itimer(seconds):
             signal.setitimer(signal.ITIMER_REAL, 0)
 
-@contextlib.contextmanager
-def timing_out_alarm(seconds):
```
Why did we remove the itimer post 2.6 stuff here?

smiklosovic replied:
@bowensong was this Windows specific?

bowensong replied:
@josh-mckenzie @smiklosovic

No, the removed method is not Windows specific.

The reason for removal:
1. the only code referenced to this method was under the `if is_win(): ... else: ...` branch below, between line 129 and 143. That chunk of code is removed because they are Windows specific.
2. Python 2.6 support was ended in 2013, nearly 9 years ago. This code has been practically sitting in an unreachable branch since then.
3. the comment says it was kept for potentially faster tests, but the there's no test using it.

Any reason to keep the unused and unreachable code in the there? I can add it back if this is a blocker.

smiklosovic replied:
thanks @bowensong , I think that the removal is justifiable.

jmckenzie-dev commented on pylib/cqlshlib/test/test_cqlsh_output.py:
```diff
@@ -587,10 +584,7 @@ def test_prompt(self):
 
-            start_index = 0
-            if c.realtty:
-                self.assertEqual(outputlines[start_index], 'use NONEXISTENTKEYSPACE;')
```
Why did we remove the assertion check for index 0?

smiklosovic replied:
moved back.

jmckenzie-dev commented on pylib/cqlshlib/copyutil.py:
```diff
@@ -884,7 +884,7 @@ def __init__(self, fname, options):
         self.skip_rows = options.copy['skiprows']
         self.fname = fname
-        self.sources = None  # must be created later due to pickle problems on Windows
+        self.sources = None
```
Without Windows' issues, should we instead initialize this at declaration time now?

smiklosovic replied:
looking into this, this one seems to be easy to move to constructor.

jmckenzie-dev commented on pylib/cqlshlib/copyutil.py:
```diff
@@ -1299,9 +1299,9 @@ def __init__(self, inpipe, outpipe, worker_pipes, fname, options):
-        self.inmsg = None  # must be created after forking on Windows
-        self.outmsg = None  # must be created after forking on Windows
-        self.worker_channels = None  # must be created after forking on Windows
+        self.inmsg = None
```
Same here: without Windows' issues, should we instead initialize these at declaration time now?

jmckenzie-dev commented on pylib/cqlshlib/copyutil.py:
```diff
@@ -1405,8 +1405,8 @@ def __init__(self, params, target):
         self.outpipe = params['outpipe']
-        self.inmsg = None  # must be initialized after fork on Windows
-        self.outmsg = None  # must be initialized after fork on Windows
+        self.inmsg = None
```
See above ☝️ . No reason not to init at decl?

smiklosovic replied:
I am not sure about this, I admit I know 0 about threading in Python. But by mere looking at it, I do not think this is a big deal anyway. Maybe @bowensong  may shed some light on this? But I would personaly not touch it and just moved along.

bowensong replied:
I believe that we should move the code from `on_fork()` to `__init__()`, and may also remove the `on_fork()` method if it's no longer used.

The problem with Windows and Python `multiprocessing` is that Windows doesn't support `fork()`, therefore Python implemented a workaround. On Windows, Python `multiprocessing` library uses `pickle` to serialise everything in memory, spawn a new process, and then restores the memory content from the serialised data. The `ReceivingChannel` and `SendingChannel` objects are not serialisable because they have file descriptors (which I believe it's called a file handle on Windows) in them, therefore the code has to handle them after the fake `fork()`.

However, I'm concerned that moving the code from `on_fork()` to `__init__()` may have other unintended side effects. For example, the file descriptors (FDs) will be opened before the fork, in some edge cases the fork may never happen (e.g.: an exception raised in or just after the init, but before the fork). Where's the code responsible for closing the FDs on the parent process side? Will this cause any FD leak? This clearly requires further work to find out.

To be honest, I don't think removing the comments without addressing the above is a wise move. Future developers wouldn't have the opportunity to understand why the code is written in this way if the comment is removed.

smiklosovic replied:
hi @bowensong , thanks for the very detailed answer. I can revert the removal of the comment or rather change it in such a way it would mention a ticket the work (or rather investigation into this matter) would be addressed in. I honestly think the refactoring of this would prolong / delay the merging of this branch unncessarilly.

bowensong replied:
@smiklosovic I agree. Refactoring this shouldn't be a part of this as it will unnecessarily prolong / delay this PR. The work is ideally done in a separate PR. I personally would prefer a comment added here and make it clear why the code was written in this way, and link to a JIRA ticket for refactoring this. Or, at minimal, a comment saying the code was written in this way for Windows `fork()` compatibility and is no longer necessary.

jmckenzie-dev commented on test/conf/cdc.yaml:
```diff
@@ -1,4 +1 @@
 cdc_enabled: true
-# Compression enabled since uncompressed + cdc isn't compatible w/Windows
-commitlog_compression:
```
What do we think is the most common config? I'd think compression on and probably lz4 because it's legit. If so, we should test cdc w/the expected common parameters. i.e. I'd leave this here on that grounds.

smiklosovic replied:
"default" in cassandra.yaml is uncompressed. I doubt people are using compressed commit logs more often as it is not default.

jmckenzie-dev commented on test/unit/org/apache/cassandra/io/sstable/SSTableLoaderTest.java:
```diff
@@ -101,6 +101,8 @@ public void cleanup()
               We force a GC here to force buffer deallocation, and then try deleting the directory again.
               For more information, see: http://bugs.java.com/bugdatabase/view_bug.do?bug_id=4715154
               If this is not the problem, the exception will be rethrown anyway.
+
```
I'd massage this message a bit to remove the "Windows does not allow" etc. etc. completely.

jmckenzie-dev commented:
Some other changes I recommend:
* In CommitLogSegmentManagerCDC, remove the Arrays.stream slower on Windows comment
* In CompactionTask:175, remove the comment about ifile and dfile on Windows
* In PathUtils.java, remove the comments about Windows being picky about renaming

smiklosovic committed:
Remove Windows-specific classes and related code

patch by Stefan Miklosovic; reviewed by Joshua McKenzie, Bowen Song, Berenguer Blasi for CASSANDRA-16956
